File: 33_Search_in_Rotated_Sorted_Array.cpp
/*
Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.
(i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).
You are given a target value to search. If found in the array return its index, otherwise return -1.
You may assume no duplicate exists in the array.
Your algorithm's runtime complexity must be in the order of O(log n).

Example 1:
Input: nums = [4,5,6,7,0,1,2], target = 0
Output: 4

Example 2:
Input: nums = [4,5,6,7,0,1,2], target = 3
Output: -1


[4,5,6,7,8,1,2,3]
8
*/
#include <iostream>
#include <vector>

using namespace std;

class Solution {

    public:
        int search(vector<int>& nums, int target) {
            int len = nums.size();
            if(len == 0) return -1;
            int minIndex = getMinIndex(nums);
            if(target == nums[minIndex]) return minIndex; 
            int front = 0, end = len - 1, mid;
            if(target <= nums[end]) front = minIndex; 
            else end = minIndex - 1;
            while(front <= end) {
                mid = front + (end - front) / 2;
                if(target < nums[mid]) {
                    end = mid - 1;
                } else if(target > nums[mid]) {
                    front = mid + 1;
                } else {
                    return mid;
                }
            }
            return -1;
        }

        // find the min value index
        int getMinIndex(vector<int>& nums) {
            int front = 0, end = nums.size() - 1, mid;
            while(front < end) {
                mid = front + (end - front) / 2;
                if(nums[mid] > nums[end]) front = mid + 1;
                else end = mid;
            }
            return front;
        }
                
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/33_Search_in_Rotated_Sorted_Array.py
"""
Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.
(i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).
You are given a target value to search. If found in the array return its index, otherwise return -1.
You may assume no duplicate exists in the array.
Your algorithm's runtime complexity must be in the order of O(log n).

Example 1:
Input: nums = [4,5,6,7,0,1,2], target = 0
Output: 4

Example 2:
Input: nums = [4,5,6,7,0,1,2], target = 3
Output: -1


[4,5,6,7,8,1,2,3]
8
"""
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r, mid = 0, len(nums) - 1, None
        if r == -1:
            return -1
        while l < r:
            mid = l + (r - l) // 2
            if nums[mid] > nums[r]:
                l = mid + 1
            else:
                r = mid
        minIndex = l
        front, end = 0, len(nums) - 1
        if nums[minIndex] == target:
            return minIndex
        elif target > nums[end]:
            end = minIndex - 1
        else:
            front = minIndex
        while front <= end:
            mid = front + (end - front) // 2
            if nums[mid] < target:
                front = mid + 1
            elif nums[mid] > target:
                end = mid - 1
            else:
                return mid
        return -1
            